client/src/wayland_obj/layer_shell.rs

```diff
@@ -0,0 +1,168 @@
+use std::{cell::RefCell, rc::Rc};
+
+use wayland_client::{protocol::wl_buffer::WlBuffer, GlobalImplementor, NewProxy, Proxy};
+
+pub(super) use wayland_protocols::wlr::unstable::layer_shell::v1::client::{
+    zwlr_layer_shell_v1::{Layer, ZwlrLayerShellV1 as WlrLayerShell},
+    zwlr_layer_surface_v1::{self as wlr_layer_surface, ZwlrLayerSurfaceV1 as WlrLayerSurface}
+};
+
+use crate::area::{Origin, Size};
+
+use super::{Buffer, Output, Surface};
+
+/// The minimum version of the wl_layer_shell protocol to bind to.
+pub(super) const WLR_LAYER_SHELL_VERSION: u32 = 1;
+
+/// LayerShell
+pub(super) struct LayerShellManager {
+    shell: Option<WlrLayerShell>
+}
+
+impl LayerShellManager {
+    pub fn new() -> Self {
+        LayerShellManager { shell: None }
+    }
+
+    pub fn create_layer_surface(
+        &self,
+        wl_surface: Surface,
+        output: Option<&Output>,
+        layer: Layer
+    ) -> Result<LayerSurface, ()> {
+        use wayland_client::Interface; // for ::NAME
+        trace!("create_layer_surface");
+        let shell = self.shell.as_ref().expect(&format!(
+            "No LayerShell registered. Make sure your compositor supports the {} protocol",
+            WlrLayerSurface::NAME
+        ));
+
+        let layer_surface = shell.get_layer_surface(
+            &wl_surface,
+            output.map(AsRef::as_ref),
+            layer,
+            Default::default(),
```

Comment:
Please prefer `StructName::default`, as it makes it clearer what is actually being defaulted.